kernel/src/ipc.rs

```diff
@@ -0,0 +1,148 @@
+//! the syscall interface are buttons.
+
+use ::{AppId, AppSlice, Container, Callback, Driver, Shared};
+use ::process;
+
+struct IPCData {
+    shared_memory: [Option<AppSlice<Shared, u8>>; 8],
+    client_callbacks: [Option<Callback>; 8],
+    callback: Option<Callback>,
+}
+
+impl Default for IPCData {
+    fn default() -> IPCData {
+        IPCData {
+            shared_memory: [None, None, None, None, None, None, None, None],
+            client_callbacks: [None, None, None, None, None, None, None, None],
+            callback: None,
+        }
+    }
+}
+
+pub struct IPC {
+    data: Container<IPCData>,
+}
+
+impl IPC {
+    pub unsafe fn new() -> IPC {
+        IPC { data: Container::create() }
+    }
+
+    pub unsafe fn schedule_callback(&self,
+                                    appid: AppId,
+                                    otherapp: AppId,
+                                    cb_type: process::IPCType) {
+        self.data
+            .enter(appid, |mydata, _| {
+                let callback = match cb_type {
+                    process::IPCType::Service => mydata.callback,
+                    process::IPCType::Client => {
+                        *mydata.client_callbacks.get(otherapp.idx()).unwrap_or(&None)
+                    }
+                };
+                callback.map(|mut callback| {
+                        self.data
+                            .enter(otherapp, |otherdata, _| {
+                                if appid.idx() >= otherdata.shared_memory.len() {
+                                    return;
+                                }
+                                match otherdata.shared_memory[appid.idx()] {
+                                    Some(ref slice) => {
+                                        slice.expose_to(appid);
+                                        callback.schedule(otherapp.idx() + 1,
+                                                          slice.len(),
+                                                          slice.ptr() as usize);
+                                    }
+                                    None => {
+                                        callback.schedule(appid.idx() + 1, 0, 0);
+                                    }
+                                }
+                            })
+                            .unwrap_or(());
+                    })
+                    .unwrap_or(());
+            })
+            .unwrap_or(());
+    }
+}
+
+impl Driver for IPC {
+    fn subscribe(&self, subscribe_num: usize, callback: Callback) -> isize {
+        match subscribe_num {
+            0 /* Service callback */ => {
+                self.data.enter(callback.app_id(), |data, _| {
+                    data.callback = Some(callback);
+                    0
+                }).unwrap_or(-2)
+            }
+            svc_id /* Client callback */ => {
+                if svc_id - 1 >= 8 {
```

Comment:
It's the length of the slice that can store callbacks (the module assumes maximum 8 processing s until I implement vectors with grant memory)